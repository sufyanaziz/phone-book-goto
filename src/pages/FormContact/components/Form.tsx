import React, { useEffect, useMemo, useState } from "react";
import { useNavigate } from "react-router-dom";
import Text from "@common/components/Text";
import { css } from "@emotion/react";
import { AddSquareIcon, MinusSquareIcon } from "@assets/icon";
import Button from "@common/components/Button";
import useFormContact from "../hooks/useFormContact";
import Input from "@common/components/Input";
import HeaderTitle from "@common/components/HeaderTitle";

const formContactStyle = () => {
  return css({
    width: "100%",
    ".field-input": {
      display: "flex",
      gap: "4px",
      width: "100%",
      ".input-container": {
        width: "100%",
        marginRight: "12px",
        ":last-child": {
          marginRight: 0,
        },
        ".input": {
          width: "100%",
          marginTop: "4px",
        },
      },
    },
    ".field-phone-numbers": {
      marginTop: "12px",
      ".phone-container": {
        display: "flex",
        marginBottom: "8px",
        width: "100%",
        "&:last-child": {
          marginBottom: "0px",
        },
        ".icons": {
          display: "flex",
          marginLeft: "8px",
          ".add-icon, .minus-icon": {
            display: "flex",
            alignItems: "center",
          },
        },
      },
    },
  });
};

const Form = () => {
  const navigate = useNavigate();
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [phoneNumbers, setPhoneNumbers] = useState<string[]>([""]);

  const { onAddContact, onEditContact, result, contact } = useFormContact();

  const isDisabled = useMemo(() => {
    if (result.loading) return true;
    if (!firstName) return true;
    if (!lastName) return true;
    if (phoneNumbers.some((phone) => !phone)) return true;
    return false;
  }, [firstName, lastName, phoneNumbers, result.loading]);

  const resetState = () => {
    setFirstName("");
    setLastName("");
    setPhoneNumbers([""]);
  };

  useEffect(() => {
    if (!result.loading && result.isSuccess) {
      navigate("/");
      resetState();
    }
  }, [result.isSuccess, result.loading]);

  useEffect(() => {
    if (contact.isHaveContact) {
      const { contact_by_pk } = contact.data;
      setFirstName(contact_by_pk?.first_name);
      setLastName(contact_by_pk?.last_name);
      setPhoneNumbers(
        contact_by_pk?.phones?.map((phone: { number: string }) => phone.number)
      );
    }
    return () => resetState();
  }, [contact.data]);

  return (
    <div css={formContactStyle}>
      <HeaderTitle
        title={contact.isHaveContact ? "Edit Form Contact" : "Form Contact"}
        css={{ display: "flex", justifyContent: "center" }}
      />
      <div className="field-input">
        <div className="input-container">
          <Text text="First Name" />
          <Input
            className="input"
            placeholder="Input First Name..."
            acceptValue="string-number"
            value={firstName}
            onChange={setFirstName}
          />
        </div>
        <div className="input-container">
          <Text text="Last Name" />
          <Input
            className="input"
            placeholder="Input Last Name..."
            acceptValue="string-number"
            value={lastName}
            onChange={setLastName}
          />
        </div>
      </div>
      <div className="field-phone-numbers">
        <Text text="Phone Numbers" css={{ marginBottom: "4px" }} />
        {phoneNumbers.map((phone, listIdx) => {
          return (
            <div className="phone-container" key={`phone-${listIdx}`}>
              <Input
                className="input"
                placeholder={`Phone ${listIdx + 1}`}
                value={phone}
                acceptValue="number-only"
                onChange={(value) => {
                  const phones = [...phoneNumbers];
                  phones[listIdx] = value;
                  setPhoneNumbers(phones);
                }}
                disabled={contact.isHaveContact}
              />
              {!contact.isHaveContact ? (
                <div className="icons">
                  {phoneNumbers.length > 1 ? (
                    <div
                      className="minus-icon"
                      onClick={() => {
                        setPhoneNumbers(
                          phoneNumbers.filter((_, idx) => idx !== listIdx)
                        );
                      }}
                      css={{
                        cursor: contact.isHaveContact ? "not-allowed" : "unset",
                      }}
                    >
                      <MinusSquareIcon />
                    </div>
                  ) : null}
                  <div
                    className="add-icon"
                    onClick={() => {
                      setPhoneNumbers((prev) => [...prev, ""]);
                    }}
                    css={{
                      cursor: contact.isHaveContact ? "not-allowed" : "unset",
                    }}
                  >
                    <AddSquareIcon />
                  </div>
                </div>
              ) : null}
            </div>
          );
        })}
      </div>
      <div css={{ marginTop: "16px", width: "100%" }}>
        <Button
          label={!contact.isHaveContact ? "Add Contact" : "Edit Contact"}
          disabled={isDisabled}
          onClick={() => {
            if (!contact.isHaveContact) {
              onAddContact(firstName, lastName, phoneNumbers);
            } else {
              onEditContact(firstName, lastName);
            }
          }}
        />
      </div>
    </div>
  );
};

export default Form;
