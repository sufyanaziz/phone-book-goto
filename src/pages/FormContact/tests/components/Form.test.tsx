import React from "react";
import renderer from "react-test-renderer";
import Form from "@pages/FormContact/components/Form";
import { BrowserRouter as Router } from "react-router-dom";
import { ApolloClient, ApolloProvider, InMemoryCache } from "@apollo/client";

const client = new ApolloClient({
  uri: "",
  cache: new InMemoryCache(),
});

describe("Component FormContact", () => {
  it("should render FormContact component correctly", () => {
    const component = renderer.create(
      <ApolloProvider client={client}>
        <Router>
          <Form />
        </Router>
      </ApolloProvider>
    );
    expect(component).toMatchSnapshot();
  });
});
