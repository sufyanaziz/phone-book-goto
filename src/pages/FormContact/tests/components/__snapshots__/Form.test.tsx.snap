// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component FormContact should render FormContact component correctly 1`] = `
<div
  css={[Function]}
>
  <div
    css={[Function]}
  >
    <div
      className=""
      css={[Function]}
    >
      Form Contact
    </div>
  </div>
  <div
    className="field-input"
  >
    <div
      className="input-container"
    >
      <div
        className=""
        css={[Function]}
      >
        First Name
      </div>
      <input
        className="input"
        disabled={false}
        onChange={[Function]}
        placeholder="Input First Name..."
        value=""
      />
    </div>
    <div
      className="input-container"
    >
      <div
        className=""
        css={[Function]}
      >
        Last Name
      </div>
      <input
        className="input"
        disabled={false}
        onChange={[Function]}
        placeholder="Input Last Name..."
        value=""
      />
    </div>
  </div>
  <div
    className="field-phone-numbers"
  >
    <div
      className=""
      css={[Function]}
    >
      Phone Numbers
    </div>
    <div
      className="phone-container"
    >
      <input
        className="input"
        disabled={false}
        onChange={[Function]}
        placeholder="Phone 1"
        value=""
      />
      <div
        className="icons"
      >
        <div
          className="add-icon"
          css={
            {
              "cursor": "unset",
            }
          }
          onClick={[Function]}
        >
          <svg
            fill="none"
            height="24px"
            viewBox="0 0 24 24"
            width="23px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 12H15"
              stroke="#435334"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
            />
            <path
              d="M12 9L12 15"
              stroke="#435334"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
            />
            <path
              d="M3 12C3 4.5885 4.5885 3 12 3C19.4115 3 21 4.5885 21 12C21 19.4115 19.4115 21 12 21C4.5885 21 3 19.4115 3 12Z"
              stroke="#435334"
              strokeWidth="2"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    css={
      {
        "marginTop": "16px",
        "width": "100%",
      }
    }
  >
    <button
      className="primary"
      css={[Function]}
      disabled={true}
      onClick={[Function]}
    >
      Add Contact
    </button>
  </div>
</div>
`;
